src/postgraphile/http/subscriptions.ts

```diff
@@ -186,134 +191,262 @@ export async function enhanceHttpServerWithSubscriptions<
+              hookedArgs.document,
+              staticValidationRules,
             );
-            error['errors'] = validationErrors;
-            return Promise.reject(error);
-          }
-        }
+            if (validationErrors.length) {
+              return validationErrors;
+            }
+
+            // You are strongly encouraged to use
+            // `postgraphile:validationRules:static` if possible - you should
+            // only use this one if you need access to variables.
+            const { req, res } = await reqResFromSocket(ctx.socket);
+            const validationRules = pluginHook('postgraphile:validationRules', [], {
+              options,
+              req,
+              res,
+              variables: hookedArgs.variableValues,
+              operationName: hookedArgs.operationName,
+              // no meta because validation errors returned from here will be
+              // served through the error message. it contains just the GraphQLErrors
+              // (there is no result to add the meta to)
+            });
+            if (validationRules.length) {
+              validationErrors = validate(hookedArgs.schema, hookedArgs.document, validationRules);
+              if (validationErrors.length) {
+                return validationErrors;
+              }
+            }
 
-        return finalParams;
-      },
-      onOperationComplete(socket: WebSocket, opId: string) {
-        releaseContextForSocketAndOpId(socket, opId);
-      },
-
-      /*
-       * Heroku times out after 55s:
-       *   https://devcenter.heroku.com/articles/error-codes#h15-idle-connection
-       *
-       * The subscriptions-transport-ws client times out by default 30s after last keepalive:
-       *   https://github.com/apollographql/subscriptions-transport-ws/blob/52758bfba6190169a28078ecbafd2e457a2ff7a8/src/defaults.ts#L1
-       *
-       * GraphQL Playground times out after 20s:
-       *   https://github.com/prisma/graphql-playground/blob/fa91e1b6d0488e6b5563d8b472682fe728ee0431/packages/graphql-playground-react/src/state/sessions/fetchingSagas.ts#L81
-       *
-       * Pick a number under these ceilings.
-       */
-      keepAlive: 15000,
-
-      ...subscriptionServerOptions,
-    },
-    wss,
-  );
+            return hookedArgs;
+          },
+          async onError(ctx, _msg, errors) {
+            // errors returned from onSubscribe
+            const { req, res } = await reqResFromSocket(ctx.socket);
+            return handleErrors(errors, req, res);
+          },
+          async onNext(ctx, _msg, _args, result) {
+            if (result.errors) {
+              // operation execution errors
+              const { req, res } = await reqResFromSocket(ctx.socket);
+              result.errors = handleErrors(result.errors, req, res);
+              return result;
+            }
+          },
+          onComplete({ socket }, msg) {
+            releaseContextForSocketAndOpId(socket, msg.id);
```

Comment:
We need this to happen when the socket disconnects unexpectedly too, I think. E.g. if the client closes the connection/times out/etc.

Reply:
Ah yes; because of the way `graphql-ws` is built, `onComplete` will be called on abrupt closures (from any side) too. [I had that confirmed with this test, just in case.](https://github.com/enisdenjo/graphql-ws/commit/3f5c42e3deb7d014258417d20df29ef9e1fb1ed7)

The only addition I had is https://github.com/graphile/postgraphile/pull/1338/commits/6b3c5a84e2d6e497da77d110c9b3bedc35f1390a. We should be tight now, the only 2 exits for an observable are `error` and `complete`.

Reply:
Great :raised_hands: